### PR TITLE
[#17] Add validator logic for events according to schema

### PR DESCRIPTION
**Pull request by enmachscb**
### Description
Add validation of event body according to event_schema associated

### Example used for development
Since for now, we need a simpler way to validate schema, structure used for this only takes care of required validation, and maximum of one level nested object, like below:

```
{
    "data"=> ["order_id", "order_name"],
    "recipient"=>["email"]
}
```

### TODO:
If you have some TODOs, please label this PR as 'in progress' and add them
here for others to follow:
- [x] Discuss if changes fit to needs;

### Release plan (Not required)
If there are some taks that needs to be done before/after deployment
- No release plan

### Rollback plan (Not required)
If there is an error, how to rollback to the previous version

Rollback last migrations: 
- No rollback plan

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -43,6 +43,7 @@ Metrics/BlockLength:
     - 'spec/requests/**/*'
     - 'spec/services/**/*'
     - 'spec/controllers/**/*'
+    - 'spec/factories/**/*'
```
Comment (gbrlmrllo) [inline .rubocop.yml:46]: 👍

Comment (gbrlmrllo) [inline .rubocop.yml:46]: 👍

File: app/models/schemas/validator.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      @body = body
+      @schema = schema.transform_keys(&:to_s)
+    end
+
+    def valid?
+      schemer.valid?(body)
+    end
+
+    def error_message
+      "invalid-body"
```
Comment (gbrlmrllo) [inline app/models/schemas/validator.rb]: why not?

`schemer.validate(body).inspect`

File: app/controllers/v1/events_controller.rb [hunk 3]
```diff
@@ -8,14 +8,18 @@ class EventsController < ApplicationController
...
-        body: params,
+        body: event_params,
         raw_data: request.raw_post
       )
       render status: :created
     end
 
     private
 
+    def event_params
```
Comment (gbrlmrllo) [inline app/controllers/v1/events_controller.rb:19]: 👍

File: spec/models/event_spec.rb [hunk 4]
```diff
@@ -11,4 +11,39 @@
     it { is_expected.to validate_presence_of(:body) }
     it { is_expected.to validate_presence_of(:raw_data) }
   end
+
+  describe "callbacks" do
+    describe "#valid_schema_event?" do
```
Comment (gbrlmrllo) [inline spec/models/event_spec.rb]: This test must be outside the scope of the callback.

In the callbacks maybe you should test that `valid_schema_event` will be called.

```ruby
describe "callbacks" do
  describe "before save" do
    it "triggers :valid_schema_event?" do
      expect(event).to receive(:valid_schema_event?)
      event.save
    end
  end
end
```

File: spec/models/schemas/validator_spec.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe Schemas::Validator do
+  let(:event_schema) { build(:event_schema) }
+  let(:event) { build(:event, event_schema: event_schema) }
+  let(:body) { event.body }
```
Comment (gbrlmrllo) [inline spec/models/schemas/validator_spec.rb]: Mocking this data from factories leaves us a bit blind.

We don't need to create the data from a factory object, I think it is better to be explicit and put the body and the direct schema here.

File: spec/factories/event_schemas.rb [hunk 6]
```diff
@@ -4,7 +4,32 @@
   factory :event_schema do
     name { FFaker::Name.name }
     description { FFaker::Lorem.phrase }
-    schema { EventSchema.schema_structure }
+    schema do
```
Comment (gbrlmrllo) [inline spec/factories/event_schemas.rb:7]: 👍

File: spec/factories/events.rb [hunk 7]
```diff
@@ -2,8 +2,18 @@
 
 FactoryBot.define do
   factory :event do
-    body { { id: 1 } }
-    raw_data { "{}" }
+    transient do
```
Comment (gbrlmrllo) [inline spec/factories/events.rb:5]: 👍

**Review by gbrlmrllo — CHANGES_REQUESTED** [review]
Awesome job!! @enmachs code wise. 🥇 

I left a few minor comments. 💬

File: app/models/event.rb [hunk 8]
```diff
@@ -4,4 +4,24 @@ class Event < ApplicationRecord
   belongs_to :event_schema, optional: false
 
   validates :body, :raw_data, presence: true
+
+  before_save :valid_schema_event?
+
+  delegate :schema, to: :event_schema
+
+  def valid_schema_event?
```
Comment (gbrlmrllo) [inline app/models/event.rb]: I was thinking about calling this method `valid_body?`, It's actually what is being validated.

File: app/models/schemas/validator.rb (hunk 2 above)
Comment (enmachs) [reply app/models/schemas/validator.rb]: Oh will try it out, thanks!

File: spec/models/schemas/validator_spec.rb (hunk 5 above)
Comment (enmachs) [reply spec/models/schemas/validator_spec.rb]: Code will be twice, is there no problem with it? @gbrlmrllo

Comment (enmachs) [reply spec/models/schemas/validator_spec.rb]: But see the point, is understandable sacrifice some DRYness for readability

File: app/models/schemas/validator.rb (hunk 2 above)
Comment (enmachs) [reply app/models/schemas/validator.rb]: would be something like
`"#<Enumerator: #<JSONSchemer::Schema::Draft7:0x0000557bec5efb38 @root={\"type\"=>\"object\", \"properties\"=>{\"data\"=>{\"type\"=>\"object\", \"properties\"=>{\"order_id\"=>{\"type\"=>\"string\"}, \"order_name\"=>{\"type\"=>\"string\"}}, \"additionalProperties\"=>false}, \"name\"=>{\"type\"=>\"string\"}, \"recipients\"=>{\"type\"=>\"array\", \"items\"=>{\"type\"=>\"object\", \"properties\"=>{\"email\"=>{\"type\"=>\"string\"}}, \"additionalProperties\"=>false}}}, \"additionalProperties\"=>false}, @format=true, @insert_property_defaults=false, @formats=nil, @keywords=nil, @ref_resolver=#<Proc:0x0000557beb5a1d80@/home/enmachs/.rbenv/versions/2.6.6/lib/ruby/gems/2.6.0/gems/json_schemer-0.2.11/lib/json_schemer/schema/base.rb:21>>:validate_instance(#<struct JSONSchemer::Schema::Base::Instance data={\"data\"=>{\"order_id\"=>\"1\", \"order_name\"=>\"Play Station 6 Deluxe edition\"}, \"name\"=>\"RegularSchema\", \"recipients\"=>[{\"email\"=>\"admin3@email.com\"}]}, data_pointer=\"\", schema={\"type\"=>\"object\", \"properties\"=>{\"data\"=>{\"type\"=>\"object\", \"properties\"=>{\"order_id\"=>{\"type\"=>\"string\"}, \"order_name\"=>{\"type\"=>\"string\"}}, \"additionalProperties\"=>false}, \"name\"=>{\"type\"=>\"string\"}, \"recipients\"=>{\"type\"=>\"array\", \"items\"=>{\"type\"=>\"object\", \"properties\"=>{\"email\"=>{\"type\"=>\"string\"}}, \"additionalProperties\"=>false}}}, \"additionalProperties\"=>false}, schema_pointer=\"\", parent_uri=nil, insert_property_defaults=false>)>"
`

File: app/models/event.rb (hunk 8 above)
Comment (enmachs) [reply app/models/event.rb]: done!

File: app/models/schemas/validator.rb (hunk 2 above)
Comment (enmachs) [reply app/models/schemas/validator.rb]: done!

File: spec/models/schemas/validator_spec.rb (hunk 5 above)
Comment (enmachs) [reply spec/models/schemas/validator_spec.rb]: done!

File: spec/models/schemas/validator_spec.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe Schemas::Validator do
+  # let(:event_schema) { build(:event_schema) }
+  # let(:event) { build(:event, event_schema: event_schema) }
```
Comment (gbrlmrllo) [inline spec/models/schemas/validator_spec.rb]: Can you remove these comments?

**Review by gbrlmrllo — APPROVED** [review]
Awesome! @enmachs Thanks for this great job.

A last comment, and merge when you want! 🚢

File: spec/models/schemas/validator_spec.rb (hunk 9 above)
Comment (enmachscb) [reply spec/models/schemas/validator_spec.rb]: lol sure will do! my bad

**Comment by enmachscb** [comment]
Merging now! @gbrlmrllo

**Pull request merged**